### PR TITLE
monitor: Reduce number of buffers and objects created

### PR DESCRIPTION
**Pull request by rlenglet**

### CODE REVIEW

**Comment by rlenglet** [comment]
Added benchmarks. Run inside the cilium-master Vagrant VM:

```
go test -v github.com/cilium/cilium/monitor/payload -bench . -check.vv -check.b
=== RUN   Test
START: monitor_payload_test.go:102: PayloadSuite.BenchmarkReadMetaPayload
PASS: monitor_payload_test.go:102: PayloadSuite.BenchmarkReadMetaPayload	   50000	     26181 ns/op

START: monitor_payload_test.go:81: PayloadSuite.BenchmarkWriteMetaPayload
PASS: monitor_payload_test.go:81: PayloadSuite.BenchmarkWriteMetaPayload	  500000	      4450 ns/op

OK: 2 passed
--- PASS: Test (3.92s)
PASS
ok  	github.com/cilium/cilium/monitor/payload	3.927s
```

This indicates that we can read ~38k and write ~224k samples/sec.

**Comment by rlenglet** [comment]
CI tests pass except the "K8s multi node Tests".

File: cilium/cmd/monitor.go [hunk 1]
```diff
@@ -257,35 +256,20 @@ start:
...
-		if _, err := io.ReadFull(conn, metaBuf); err != nil {
+		if err := payload.ReadMetaPayload(conn, &meta, &pl); err != nil {
 			if err == io.EOF {
+				// EOF may be due to invalid payload size. Close the connection just in case.
+				conn.Close()
+				log.WithError(err).Warn("connection closed")
 				time.Sleep(connTimeout)
 				goto start
+			} else {
+				log.WithError(err).Fatal("decoding error")
```
Comment (tgraf) [inline cilium/cmd/monitor.go:269]: I assume `io.ErrUnexpectedEOF` is intentionally not handled above and treated as a fatal error like all other errors. Thus requiring a manual restart of the monitor. I think that's fine.

**Review by tgraf — APPROVED** [review]

File: cilium/cmd/monitor.go (hunk 1 above)
Comment (rlenglet) [reply cilium/cmd/monitor.go:269]: I didn't think of this error. It seems like I should handle it like EOF above. I'll update the PR.

**Review by aalemayhu — APPROVED** [review]
Looks good, :+1: for the test.

File: cilium/cmd/monitor.go (hunk 1 above)
Comment (rlenglet) [reply cilium/cmd/monitor.go:269]: Done.

**Pull request merged**